remove: make `TarUtils` final and clean up internal methods

ppkarwasz opened the pull request:
`TarUtils` already had a private constructor and was only used within the `o.a.c.compress.archivers.tar` package. This PR makes the class explicitly final and simplifies its internal API:

* Mark `TarUtils` as `final`.
* Change `protected` methods to package-private (they were never usable outside the package due to the private constructor).
* Remove deprecated non-`public` methods.
* Simplify `parsePaxHeaders`:

  * Require a non-negative `headerSize`. The special value `-1` (“unlimited”) was only used in tests.
  * Update tests to supply a valid `headerSize`.
  * Standardize error handling for invalid PAX 0.0 sparse records: all parsing errors now throw `ArchiveException` (previously one case threw `IOException`).

A changelog entry is included even though these are internal changes, to give users a reference point in case any unintended side effects arise.

This change is linked with #710, which needs to add a parameter to the currently protected `parsePaxHeaders` method.

garydgregory commented on pom.xml:
```diff
@@ -300,7 +300,26 @@ Brotli, Zstandard and ar, cpio, jar, tar, zip, dump, 7z, arj.
+                ~
+                ~ TODO: Remove after release 1.29.0.
+                -->
+              <overrideCompatibilityChangeParameters>
```
Obvious -1: We can't break binary compatibility for a trivial style/design change.

garydgregory replied:
OK, that makes sense; let's see what JApiCmp says about the ticket I created: https://github.com/siom79/japicmp/issues/426 

FWIW, the JLS' definition of binary compatible for final classes has changed from version to version, which I didn't bother to talk about in the ticket.

garydgregory reviewed (commented):
Obvious -1: We can't break binary compatibility for a trivial style/design changes.

ppkarwasz commented on src/main/java/org/apache/commons/compress/archivers/tar/TarUtils.java:
```diff
@@ -48,8 +48,7 @@
  */
-// CheckStyle:HideUtilityClassConstructorCheck OFF (bc)
-public class TarUtils {
+public final class TarUtils {
```
I wouldn’t propose a change that breaks binary compatibility for users.

You’re correct that in general marking a class `final` is a binary-incompatible change, but in this specific case the rule only applies if the class **can be subclassed**. Since `TarUtils` has a **private constructor** and no nested subclasses, it cannot be extended, so making it `final` does not remove any legal use and therefore does not break compatibility.

This is a subtle edge case: tools like **BND** recognize it correctly, while **JapiCmp** currently does not.

ppkarwasz commented:
I targeted this PR to the `commons_io_2_21_0`, to limit conflicts, when that branch is merged back into `master`.

ppkarwasz commented:
To simplify the release I bumped the version of `japicmp` in this PR, overriding the value inherited from `commons-parent`.

Since the release of `commons-io` and `commons-parent` can be in parallel though, we can probably use both new version in the `commons_io_2_21_0` branch.

garydgregory commented on pom.xml:
```diff
@@ -89,6 +87,12 @@ Brotli, Zstandard and ar, cpio, jar, tar, zip, dump, 7z, arj.
     <checkstyle.config.file>${basedir}/src/conf/checkstyle/checkstyle.xml</checkstyle.config.file>
     <checkstyle.suppress.file>${basedir}/src/conf/checkstyle/checkstyle-suppressions.xml</checkstyle.suppress.file>
     <checkstyle.resourceExcludes>LICENSE.txt, NOTICE.txt, **/maven-archiver/pom.properties</checkstyle.resourceExcludes>
+    <!--
```
There is no need to make this PR noisier by moving comments and changing their format to whatever this "~" is.

ppkarwasz replied:
I moved all the overwritten versions together, so that we can more easily check each of those overrides, when we upgrade `commons-parent`.

I minimized the changes in https://github.com/apache/commons-compress/pull/712/commits/adcb46b6e785fac63153f3bb604c831d9f0bb37e

ppkarwasz commented:
@garydgregory,

In 333e11f8db47161efe69fdd56e385c9ae761dc6d, I added a Maven Enforcer rule intended to remind us to remove the `commons.japicmp.version` override with the next `commons-parent` upgrade.

Unfortunately, the rule fails on CI while passing locally, and I was unable to reproduce the CI failure despite trying:

* Using the same Maven version as CI (`3.9.11`)
* Clearing the local `~/.m2/repository` cache
* Running the build with `CI=true`
* Removing all repository caches

Since the build consistently fails only on CI, I dropped the rule in 854affde56ca12830a7a7353ce46108439e7c11b.

garydgregory commented:
PR merged, thank you @ppkarwasz 👍
